tests/scipy_optimize_test.py

```diff
@@ -0,0 +1,260 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from absl.testing import absltest
+import numpy as np
+
+import jax
+from jax import numpy as np
+from jax import test_util as jtu
+import jax.scipy.sparse.linalg
+from jax.config import config
+from jax.scipy.optimize.bfgs_minimize import BFGSResults, bfgs_minimize
+from jax.scipy.optimize.line_search import LineSearchResults, line_search
+
+config.parse_flags_with_absl()
+
+from jax.config import config
+
+config.parse_flags_with_absl()
+
+
+def line_search_nojit(value_and_gradient, position, direction, f_0=None, g_0=None, max_iterations=50, c1=1e-4, c2=0.9):
```

Comment:
Can you add a comment saying more about why these functions are re-defined here? It seems like it may be better to compare the original unjitted function with a jitted version of itself, but I may be missing something.

Reply:
These have the while_loop replaced with python while loops. Originally, much easier to be sure the algorithm was written properly. When going to while_loops with dynamic termination criteria, for jitting, I used these to compare. I don't think they are needed anymore since we can just compare with scipy.optimize.minimize. I'll remove them.

Reply:
However, there is no line_search test otherwise. There's nothing to compare to. Perhaps it's best to leave these pythonic versions to compare to?

Reply:
If you only need a function for tests, move it to the test file.

That said, there are probably some good ideas from the SciPy tests for line search:
https://github.com/scipy/scipy/blob/master/scipy/optimize/tests/test_linesearch.py

(SciPy also have a variety of BFGS tests that could be worthy of consideration)

Reply:
Good point, using scipy's line search tests would be useful. It looks extensive. I think it would make more sense to test line search in a followup PR where we could implement the same line search. Mine is not zooming. Most importantly is that my BFGS performs similarly to scipy BFGS for a suite of problems. Given that we also analytically initialise the inverse hessian it's also possible that our performs better in some cases. So, for now I'll remove line search tests and compare BFGS on more test problems.